### ISSUE TITLE
Add support for overriding the catalog url

### ISSUE THREAD

**Issue by naufraghi** [comelz/quark#9]
Sometime is useful to test a dependency in a big project, one can generate the freeze file and use that to override the dependency branch for example and avoid integrity errors in the deps resolution.

But can we add a better option and point to a custom catalog where a dependency is the tested one?

Something like:
```
$ quark up --catalog file://$HOME/project/catalog.json
```
Where `catalog.json` is:
```
{
    "dep" :{ "url":"git+ssh://github.com/comelz/dep.git#branch=dev" }
}
```
In format of a patch `original_catalog.update(local_catalog)` or a full copy of the original catalog.

**Comment by woggioni** [comment]
The original idea was that you should change your quark.json file in that
case, changing the URL/branch the dependency that you want to test. Passing
it as command line parameter is weird because it would not reflect the
actual dependency tree of the package (if you just checkout and build it
you'll still use the dependencies written in the quark.json file)

**Comment by naufraghi** [comment]
What if:
```
prj -> dep1 + dep2 + dep3
dep1 -> dep3
```
And I'd like to test the integration of `dep3#branch=ng` in `prj`.

If I change the url in `prj` only, I will have a conflict during the resolution because `dep3 != dep3#branch=ng`, so I need to make a `dep1#branch=test-dep3-ng` branch and change that url too in `prj`.

The suggested `--catalog` option, or, let's call it `--test-catalog` is just a way to avoid that kind of viral test branches. No one in production is supposed to use it.

The other (already implemented) option and manually modify and commit the lock file to achieve the same objective (avoiding viral test branches), but IMHO is quite hackish.

Did I miss some other way to cover this use case?

**Comment by woggioni** [comment]
In the original prototype of quark the algorithm that was in place was to take for every package the commit/branch/tag specified in the node nearest to the root of the dependency tree (and to fail if that was ambiguous, like two direct children depending on different versions of another package). Your use case would have been resolved with the version specified in `prj`.
I was then asked to change it to simply break in case of dependency version conflict for the sake of simplicity, so yes, presently there is the limitation that you underlined. Another (a bit hacky) solution would be generate and then manually edit the freeze file as you suggested, it actually also seems to me a quite good one, did you encounter any issue with it?

**Comment by naufraghi** [comment]
I only see two minor downsides in the lock-file way:

- it is implicit (we'd better document the hack as the suggested way to
tackle the use case)
- it a committed project file (the `--test-catalog` may appear in the CI
configuration only)

Being a CI only option can allow for example to test the new branch using a
new pipeline in any existing project, without touching the working code.

We will test the lock-file option better and resume this issue with a
better understanding if necessary.

**Comment by woggioni** [comment]
well, you would have to create a new branch to be able to use different dependencies, but it would only apply to the root package, not to all of the dependencies (as it would be the case if you edited the `quark.json` file). It is true that you cannot edit the dependency tree at the pipeline level in any way other than switching the pipeline to the development branch but I personally do not see it as a big limitation and I think it actually helps to keep things simple (you wouldn't have to check what are the CLI options that the CI is passing to quark to fully understand the build configuration)

**Comment by cvtsi2sd** [comment]
I too longed for this feature for a while when migrating projects to test a new catalog before committing it. I see no particular issue to originally proposed form besides the fact that in principle it is supported for each subprojects.quark to have its own catalog, so it's not clear what catalog exactly the override option would refer to.

An idea that comes to mind would be to allow a "long form" option to override a specific catalog URL, and a shorthand option that overrides the catalog referenced in the top-level subprojects.quark, leaving the others alone. Something like:

    quark up --catalog-override https://www.example.com/catalog.json=file://$HOME/project/catalog.json
    quark up --catalog-override file://$HOME/project/catalog.json

This would be quite easy to add to the current implementation.

**Comment by cvtsi2sd** [comment]
@naufraghi something like this: https://github.com/comelz/quark/tree/add-catalog-override

**Comment by cvtsi2sd** [comment]
Implemented in #19
